packages/yup/src/dateRange.js

```diff
@@ -8,42 +9,121 @@ dayjs.extend(customParseFormat);
 dayjs.extend(isSameOrAfter);
 dayjs.extend(isSameOrBefore);
 
-export default function(
-  { min, max, format = 'MM/DD/YYYY', start = 'startDate', end = 'endDate' },
-  msg
-) {
-  const minDate = dayjs(min, format, true);
-
-  const maxDate = dayjs(max, format, true);
-
-  // Can't use arrow function because we rely on 'this' referencing yup's internals
-  return this.test({
-    name: 'dateRange',
-    exclusive: true, // Validation errors don't stack
-    // NOTE: Intentional use of single quotes - yup will handle the string interpolation
-    message:
-      msg ||
-      `Date Range must be between ${minDate.format(
-        format
-      )} and ${maxDate.format(format)}`,
-    test(value) {
-      if (!value) return false;
-      let startDate = get(value, start);
-      let endDate = get(value, end);
-
-      if (!startDate || !endDate) return false;
-
-      startDate = dayjs(startDate, format, true);
-
-      endDate = dayjs(endDate, format, true);
-
-      return (
-        startDate.isValid() &&
-        endDate.isValid() &&
-        endDate.isSameOrAfter(startDate) &&
-        startDate.isSameOrAfter(minDate) &&
-        endDate.isSameOrBefore(maxDate)
-      );
-    },
-  });
+const defaultOptions = {
+  startKey: 'startDate',
+  endKey: 'endDate',
+  format: 'MM/DD/YYYY',
+};
+
+export default class DateRangeSchema extends yup.mixed {
+  constructor({ startKey, endKey, format } = defaultOptions) {
+    super({
+      type: 'dateRange',
+    });
+
+    this.startKey = startKey;
+    this.endKey = endKey;
+    this.format = format;
+
+    this.withMutation(() => {
+      this.transform(function mutate(value) {
+        const start = get(value, startKey);
+        const end = get(value, endKey);
+
+        if ((start && !end) || (!start && end)) {
```

Comment:
what about
```
if (!start || !end) {
```

Reply:
I wish I could tell you why I did that... 🤷‍♂